java/src/org/openqa/selenium/remote/http/RetryRequest.java

```diff
@@ -17,89 +17,55 @@
-          .handleResultIf(response -> (response).getStatus() == HTTP_UNAVAILABLE)
-          .withMaxRetries(2)
-          .onRetry(
-              e ->
-                  LOG.log(
-                      getDebugLogLevel(),
-                      "Failure due to server error #{0}. Retrying.",
-                      e.getAttemptCount()))
-          .build();
+  private static final int RETRIES_ON_CONNECTION_FAILURE = 3;
+  private static final int RETRIES_ON_SERVER_ERROR = 2;
+  private static final int NEEDED_ATTEMPTS =
+      Math.max(RETRIES_ON_CONNECTION_FAILURE, RETRIES_ON_SERVER_ERROR) + 1;
 
   @Override
   public HttpHandler apply(HttpHandler next) {
-    return req ->
-        Failsafe.with(fallback)
-            .compose(serverErrorPolicy)
-            .compose(connectionFailurePolicy)
-            .get(() -> next.execute(req));
-  }
+    return req -> {
+      for (int i = 0; i < NEEDED_ATTEMPTS; i++) {
+        HttpResponse response;
+
+        try {
+          response = next.execute(req);
+        } catch (RuntimeException ex) {
+          if (ex.getCause() instanceof ConnectException && i < RETRIES_ON_CONNECTION_FAILURE) {
+            LOG.log(LOG_LEVEL, "Retry #" + (i + 1) + " on ConnectException", ex);
+            continue;
+          }
 
-  private static HttpResponse getFallback(
-      ExecutionAttemptedEvent<? extends HttpResponse> executionAttemptedEvent) throws Exception {
-    if (executionAttemptedEvent.getLastException() != null) {
-      Exception exception = (Exception) executionAttemptedEvent.getLastException();
-      if (exception.getCause() instanceof ConnectException) {
-        return new HttpResponse()
-            .setStatus(HTTP_CLIENT_TIMEOUT)
-            .setContent(asJson(Map.of("value", Map.of("message", "Connection failure"))));
-      } else throw exception;
-    } else if (executionAttemptedEvent.getLastResult() != null) {
-      HttpResponse response = executionAttemptedEvent.getLastResult();
-      if ((response.getStatus() == HTTP_INTERNAL_ERROR
-              && Integer.parseInt(response.getHeader(HttpHeader.ContentLength.getName())) == 0)
-          || response.getStatus() == HTTP_UNAVAILABLE) {
-        return new HttpResponse()
-            .setStatus(response.getStatus())
-            .setContent(asJson(Map.of("value", Map.of("message", "Internal server error"))));
+          throw ex;
+        }
+
+        boolean isServerError =
+            (response.getStatus() == HTTP_INTERNAL_ERROR && response.getContent().length() == 0)
+                || response.getStatus() == HTTP_UNAVAILABLE;
+
+        if (isServerError && i < RETRIES_ON_SERVER_ERROR) {
+          LOG.log(LOG_LEVEL, "Retry #" + (i + 1) + " on ServerError: " + response.getStatus());
+          continue;
+        }
+
+        return response;
       }
-    }
-    return executionAttemptedEvent.getLastResult();
+
+      throw new IllegalStateException(
+          "Effective unreachable code reached, please raise a bug ticket.");
```

Comment:
This line is only reachable in case somebody does change the `static final int` values at the top of the class to inconsistent values, e.g. more retries than total attemps. So this is not related to any local networking issue.

I could replace this with an `AssertionError` to show this is a state we should never ever reach.

Reply:
Oh, I missed that. Then I would say that in the throw instead of telling the user to raise a ticket.